elfcore/src/coredump.rs

```diff
@@ -1525,3 +1632,45 @@ fn write_va_regions<T: Write>(
 
     Ok(written)
 }
+
+/// A builder for generating a core dump of a process by pid,
+/// optionally with custom notes with content from files
+pub struct CoreDumpBuilder {
+    pv: ProcessView,
+    custom_notes: Vec<CustomFileNote>,
+}
+
+impl CoreDumpBuilder {
+    /// Create a new core dump builder for the process with the provided PID
+    pub fn new(pid: libc::pid_t) -> Result<Self, CoreError> {
+        let pv = ProcessView::new(pid)?;
+        Ok(Self {
+            pv,
+            custom_notes: Vec::new(),
+        })
+    }
+
+    /// Add the contents of a file as a custom note to the core dump
+    pub fn add_custom_file_note(&mut self, name: &str, file: File, note_len: usize) -> &mut Self {
+        self.custom_notes.push(CustomFileNote {
+            name: name.to_owned(),
+            file,
+            note_len,
+        });
+        self
+    }
+
+    /// Add the contents of a file as a custom note to the core dump
+    pub fn with_custom_file_note(mut self, name: &str, file: File, note_len: usize) -> Self {
```

Comment:
We don't need both with and add. Pick the pattern that works and use that.

Reply:
Ok I just thought it would be nice to give people the option, but currently I'm only using the add function so I'll stick with that.